doc/TuningDesign.md

```diff
@@ -0,0 +1,219 @@
+
+1) Different kernels have different valid values for block size
+2) Different invocations of the same kernel can have different values for block size if somebody changes settings
+3) We don't know how much worse this gets as we move past block size
+
+So we'll do our best to mitigate the impacts of this, but for now the set of candidate values must be provided every time we request a
+value
+
+Otherwise the ideas behind the tuning system translate directly into the design and the implementation
+
+## Implementation
+
+This section describes the implementation.
+
+If you're writing a tool, you care about tool implementation.
+
+If you want tools to know about information from your application, you care about application implementation
+
+If you're a Kokkos developer, you care about the application implementation and Kokkos implementation
+
+### Tool implementation
+
+In the past, tools have responded to the [profiling hooks in Kokkos](https://github.com/kokkos/kokkos-tools/wiki/Profiling-Hooks). This effort adds to that, there are now a few more functions (note that I'm using the C names for types. In general you can replace Kokkos_Tuning_ with Kokkos::Tuning:: in C++ tools)
+
+```c++
+void kokkosp_declare_tuning_variable(const char* name, const size_t id, Kokkos_Tuning_VariableInfo info);
+```
+
+Declares a tuning variable named `name` with uniqueId `id` and all the semantic information (except candidate values) stored in `info`.
+
+```c++
+void kokkosp_declare_context_variable(const char*, const size_t, Kokkos_Tuning_VariableInfo info, Kokkos_Tuning_VariableInfo_SetOrRange);
+```
+
+This is much like declaring a tuning variable, except the candidate values of context variables are declared immediately. In cases where they aren't known, `info.valueQuantity` will be set to `kokkos_value_unbounded`. This is fairly common, Kokkos can tell you that `kernel_name` is a string, but we can't tell you what strings a user might provide.
+
+```c++
+void kokkosp_request_tuning_variable_values(
+    const size_t contextId,
+    const size_t numContextVariables, const Kokkos_Tuning_VariableValue* contextVariableValues,
+    const size_t numTuningVariables, Kokkos_Tuning_VariableValue* tuningVariableValues, Kokkos_Tuning_VariableInfo_SetOrRange* tuningVariableCandidateValues);
+```
+
+Here Kokkos is requesting the values of tuning variables, and most of the meat is here. The contextId tells us the scope across which these variables were used. Often you'll start a measurement/timer and associate it with this ID so that when the contextID ends, you can stop the timer and know how you did.
+
+The next three arguments describe the context you're tuning in. You have the number of context variables, and an array of that size containing their values.
+
+The next four arguments describe the Tuning Variables, first the number, then some default values which you overwrite to change how Kokkos will behave. Finally, you have the candidate values you can choose among.
```

Comment:
next three?

Reply:
Good call, I'll change that. Actually, I need to reread this whole document, a lot has changed since it was written